Add long vita

NieCha opened the pull request:
# What does this PR do?

Add the awesome Long_Vita model.

The referenced project links are as follows
https://huggingface.co/VITA-MLLM/Long-VITA-16K_HF
https://github.com/VITA-MLLM/Long-VITA


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
Merge branch 'main' into add_Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
add_new_model: Long_Vita

ArthurZucker reviewed (commented):
Hey! THanks for the PR 🤗 before I review this, can you make sur your modeling code is as close as possible to `Llama` for example? meaning there are no single letter variable, you use our standard API for flash attention, make sure the layers are also similar in terms of the way we write the forwards etc! 🤗

NieCha committed:
add_new_model: Long_Vita

NieCha commented:
> Hey! THanks for the PR 🤗 before I review this, can you make sur your modeling code is as close as possible to `Llama` for example? meaning there are no single letter variable, you use our standard API for flash attention, make sure the layers are also similar in terms of the way we write the forwards etc! 🤗

Yes, we referred to the implementations of LLaMA and Qwen2_VL for developing our model and will continue to conduct checks.
Additionally, after submitting our request, we encountered two issues related to automated checks: 1) An error reporting 'import torch' in the modeling files; and 2) Numerous errors unrelated to our submitted code. 

It would be very helpful if you could explain the reasons behind these issues.🤗

NieCha committed:
Merge branch 'main' into add_Long_Vita

ArthurZucker commented:
Hey! I think the simplest way to share the model first is to put it on the hub! You won't have to deal with all these issues 🤗 
otherwise the process is a bit more involved, you should use `modular` https://huggingface.co/docs/transformers/main/en/modular_transformers to make sure it's easier to review and more aligned!

NieCha commented:
> Hey! I think the simplest way to share the model first is to put it on the hub! You won't have to deal with all these issues 🤗 otherwise the process is a bit more involved, you should use `modular` https://huggingface.co/docs/transformers/main/en/modular_transformers to make sure it's easier to review and more aligned!

Thank you for your reply😊😊. The model has already been uploaded to the Hugging Face hub, and the project address is https://huggingface.co/VITA-MLLM/Long-VITA-16K_HF.

Also, how can the code be merged without addressing these issues~

ArthurZucker commented:
In order to get the model into transformers, one needs to make sure it is as close to our standard API as possible! So as I mentionned, make sure to use modular for classes that are inherited / very similar to other models. This makes the review a lot easier! 
The modeling code needs to be cleanup, and differences isolated to better understand what is this model! 
🤗 then once you think this is ready for review you can ping us 🤗

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
Merge branch 'main' into add_Long_Vita

NieCha committed:
Merge branch 'main' into add_Long_Vita

NieCha committed:
add_new_model: Long_Vita

NieCha committed:
Merge branch 'add_Long_Vita' of https://github.com/NieCha/transformers into add_Long_Vita

ArthurZucker commented on src/transformers/models/long_vita/modular_long_vita.py:
```diff
@@ -0,0 +1,800 @@
+        return self.weight * hidden_states.to(input_dtype)
+
+
+class FlashAttention(nn.Module):
```
Flash attention is supported in transformers, we have a common api / interface to handle it! Can you try to blend it into it? (see LlamaAttention! )

ArthurZucker commented on src/transformers/models/long_vita/modular_long_vita.py:
```diff
@@ -0,0 +1,800 @@
+            self.q_norm = InternRMSNorm(self.embed_dim, eps=config.layer_norm_eps)
+            self.k_norm = InternRMSNorm(self.embed_dim, eps=config.layer_norm_eps)
+
+        if self.use_flash_attn:
+            self.inner_attn = FlashAttention(attention_dropout=config.attention_dropout)
```
that is not how we usually deal with flash attention!

ArthurZucker commented on src/transformers/models/long_vita/modular_long_vita.py:
```diff
@@ -0,0 +1,800 @@
+            self.inner_attn = FlashAttention(attention_dropout=config.attention_dropout)
+        self.proj = nn.Linear(self.embed_dim, self.embed_dim)
+
+    def _naive_attn(self, x):
+        B, N, C = x.shape
+        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
+        q, k, v = qkv.unbind(0)  # make torchscript happy (cannot use tensor as tuple)
+
+        if self.qk_normalization:
+            B_, H_, N_, D_ = q.shape
+            q = self.q_norm(q.transpose(1, 2).flatten(-2, -1)).view(B_, N_, H_, D_).transpose(1, 2)
+            k = self.k_norm(k.transpose(1, 2).flatten(-2, -1)).view(B_, N_, H_, D_).transpose(1, 2)
+        
+        attn = (q * self.scale) @ k.transpose(-2, -1)
+        attn = attn.softmax(dim=-1)
+        attn = self.attn_drop(attn)
+
+        x = (attn @ v).transpose(1, 2).reshape(B, N, C)
+        x = self.proj(x)
+        x = self.proj_drop(x)
+        return x
+
+    def _flash_attn(self, x, key_padding_mask=None, need_weights=False):
+        qkv = self.qkv(x)
+        qkv = qkv.view(qkv.shape[0], qkv.shape[1], 3, self.num_heads, -1)
+
+        if self.qk_normalization:
+            q, k, v = qkv.unbind(2)
+            q = self.q_norm(q.flatten(-2, -1)).view(q.shape)
+            k = self.k_norm(k.flatten(-2, -1)).view(k.shape)
+            qkv = torch.stack([q, k, v], dim=2)
+
+        context, _ = self.inner_attn(qkv, key_padding_mask=key_padding_mask, need_weights=need_weights, causal=False)
+        outs = self.proj(context.view(context.shape[0], context.shape[1], -1))
+        outs = self.proj_drop(outs)
```
same here, we put a function outside the class for eager attentioN!

ArthurZucker reviewed (commented):
Hey! Thanks for the ping still many things to address let's standardize 😉
